openff/toolkit/topology/molecule.py

```diff
@@ -3073,25 +3075,47 @@ def partial_charges(self, charges):
         """
         if charges is None:
             self._partial_charges = None
-        elif charges.shape == (self.n_atoms,):
-            if isinstance(charges, unit.Quantity):
-                if charges.units in unit.elementary_charge.compatible_units():
-                    self._partial_charges = charges
-            if hasattr(charges, "unit"):
-                from openmm import unit as openmm_unit
-
-                if not isinstance(charges, openmm_unit.Quantity):
-                    raise IncompatibleUnitError(
-                        "Unsupported type passed to partial_charges setter. "
-                        "Found object of type {type(charges)}."
-                    )
+            return
+
+        if not hasattr(charges, "shape"):
+            raise IncompatibleTypeError(
+                "Unsupported type passed to partial_charges setter. "
+                f"Found object of type {type(charges)}."
+            )
+
+        if not charges.shape == (self.n_atoms,):
+            raise IncompatibleShapeError(
+                "Unsupported shape passed to partial_charges setter. "
+                f"Found shape {charges.shape}, expected {(self.n_atoms,)}"
+            )
+
+        if isinstance(charges, unit.Quantity):
+            if charges.units in unit.elementary_charge.compatible_units():
+                self._partial_charges = charges.astype(float)
+                return
+
+        elif hasattr(charges, "unit"):
+            from openmm import unit as openmm_unit
+
+            if not isinstance(charges, openmm_unit.Quantity):
+                raise IncompatibleUnitError(
+                    "Unsupported type passed to partial_charges setter. "
+                    f"Found object of type {type(charges)}."
+                )
 
-                elif isinstance(charges, openmm_unit.Quantity):
-                    from openff.units.openmm import from_openmm
+            else:
+                from openff.units.openmm import from_openmm
+
+                converted = from_openmm(charges)
+                if converted.units in unit.elementary_charge.compatible_units():
+                    self._partial_charges = converted.astype(float)
```

Comment:
I noticed this case doesn't have an `else:` clause, so I think it could silently fail if somebody gets lost and passes an OpenMM quantity with non-charge-like units. This is a weird corner case but you started this PR by finding a similar failure case so here would be a great time to safeguard against it.